src/util.rs

```diff
@@ -16,8 +22,69 @@ pub(crate) fn load_rustdoc_from_file(path: &Path) -> anyhow::Result<Crate> {
         .read_to_string(&mut s)
         .with_context(|| format!("Failed to read rustdoc JSON output file {:?}", path))?;
 
-    serde_json::from_str(&s)
-        .with_context(|| format!("Failed to parse rustdoc JSON output file {:?}", path))
+    match serde_json::from_str(&s) {
+        Ok(value) => Ok(value),
+        Err(_) => {
+            // Attempt to figure out the more precise reason the deserialization failed.
+            // Several possible options and their resolutions:
+            // (1) The file isn't actually a rustdoc JSON file. The user should supply a valid file.
+            // (2) The rustdoc JSON file has a version number that is too old, and isn't supported.
+            //     The user should upgrade to a newer nightly Rust version and regenerate the file.
+            // (3) The rustdoc JSON file has a version number that is too new, and isn't supported.
+            //     The user should attempt to upgrade to a newer cargo-semver-checks version
+            //     if one is already available, or open a GitHub issue otherwise.
+
+            // The error on this line is case (1).
+            let version = serde_json::from_str::<RustdocFormatVersion>(&s).with_context(|| {
+                format!(
+                    "Failed to parse the rustdoc JSON file, and could not determine its \
+                    format version. Are you sure this is a valid rustdoc JSON file? \
+                    File path: {:?}",
+                    path
+                )
+            })?;
+
+            match version.format_version.cmp(&rustdoc_types::FORMAT_VERSION) {
+                std::cmp::Ordering::Less => {
+                    // The error here is case (2).
+                    bail!(
+                        "Failed to parse rustdoc JSON file: {path:?}\n\n\
+                        This version of cargo-semver-checks requires rustdoc JSON version {0}, \
+                        but the JSON file uses the older version {1}.\n\n\
+                        Please upgrade to a Rust nightly version where rustdoc outputs \
+                        JSON format version {0}.\n\n\
+                        It's usually easiest to upgrade to the latest versions of \
+                        both cargo-semver-checks and Rust nightly, \
+                        which should always be compatible with each other.",
+                        rustdoc_types::FORMAT_VERSION,
+                        version.format_version,
+                    )
+                }
+                std::cmp::Ordering::Greater => {
+                    // The error here is case (3).
+                    bail!(
+                        "Failed to parse rustdoc JSON file: {path:?}\n\n\
```

Comment:
As a reviewer, escaped newlines make it harder to review the legibility of these error messages.

Reply:
There's a bit of tension here and I'm not sure how to navigate it:
- If I use unescaped newlines and omit the trailing `\`, then the Rust string contains 24 spaces' worth of indentation which is not removed.
- If I escape the newlines and used a trailing `\`, then there's no indentation in the final string but as you point out, the legibility is not ideal.

One hybrid option is something like this:
```rust
bail!(
    "rustdoc output format is too old (v{1}, need v{0}) for file \
    {path:?}\n\n\
    \
    note: Using a newer Rust nightly version should help. \
    The latest cargo-semver-checks and Rust nightly should always be \
    compatible with each other.",
    rustdoc_types::FORMAT_VERSION,
    version.format_version,
)
```

Is that preferable?

Reply:
Personally, I wouldn't indent it.

Getting fancy with escapes means you are artificially making the format look like the real thing which can drift apart.

I would put this as non-blocking to be clear.